### PR TITLE
024/authentication

### PR DESCRIPTION
**Pull request by MohammedFoysal**
[connects to #024](https://trello.com/c/EetfIPTb/25-us024-create-authentication-system)

### CODE REVIEW

File: KayKnows-frontend/src/app/data.service.ts [hunk 1]
```diff
@@ -53,6 +53,37 @@ export class DataService {
 
     });
   }
+  
+  login(data, callback) {
+    this.http.post('/api/login', data).subscribe(
+      data => callback(data, null),
```
Comment (micalexis) [inline KayKnows-frontend/src/app/data.service.ts]: refactor to remove use callbacks

Comment (micalexis) [reply KayKnows-frontend/src/app/data.service.ts]:
```JAVASCRIPT
this.productService.getProducts().subscribe({
      next: products => {
        this.products = products;
        this.filteredProducts = this.products;
      },
      error: err => this.errorMessage = err
    });

```

File: KayKnows-frontend/src/app/data.service.ts [hunk 2]
```diff
@@ -53,6 +53,37 @@ export class DataService {
...
+  logout() {
+    localStorage.removeItem('token');
+    localStorage.removeItem('user_id');
+    localStorage.removeItem('user_email');
+    localStorage.removeItem('user_admin');
+    localStorage.removeItem('user_full_name');
+  }
+
+  getLoggedInUser(callback) {
+    let token = 'Bearer ' + localStorage.getItem('token');
```
Comment (micalexis) [inline KayKnows-frontend/src/app/data.service.ts]:
```JAVASCRIPT
`Bearer ${localStorage.getItem('token')}`
```

File: KayKnows-frontend/src/app/login/login.component.html [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+      <div *ngIf="loginForm.get('password').errors.required" class="password-valid error">Password is required</div>
+    </div>
+
+    <div class="error" *ngIf="response">{{ response }}</div>
+
+    <button type="submit" alt="Submit Button" class="submit"><i class="material-icons md-36">arrow_right_alt</i></button>
+  </form>
+</div>
+
+<script>
```
Comment (micalexis) [inline KayKnows-frontend/src/app/login/login.component.html]: WHAT IS THIS???

File: KayKnows-frontend/src/app/login/login.component.ts [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+      Validators.required,
+      Validators.email
+    ]),
+    password: new FormControl('', [
+      Validators.required
+    ])
+   });
+
+  constructor(dataService: DataService, private route: ActivatedRoute, private router: Router) {
+    this.dataService = dataService;
```
Comment (micalexis) [inline KayKnows-frontend/src/app/login/login.component.ts]: keep consistent approach.... use private to inject

File: KayKnows-backend/utils.js [hunk 5]
```diff
@@ -0,0 +1,31 @@
+const bcrypt = require('bcrypt');
+
+exports.hashPassword = async (rawPassword) => {
+
+    const password = rawPassword
+    const saltRounds = 10;
+
+    const hashedPassword = await new Promise((resolve, reject) => {
```
[the remark this answers is not in the document]
Comment (MohammedFoysal) [reply KayKnows-backend/utils.js]: tried to change thing, but error handling is an issue, which will need more investigation

**Pull request merged**